### PR TITLE
fix deadlock between `require_lock` and package callbacks

### PR DESCRIPTION
**Pull request by JeffBezanson**
Fix the deadlock described in https://github.com/JuliaLang/julia/pull/43339#issuecomment-990222989

Also fix one place where we were acquiring require_lock twice (which is ok since it's reentrant, but not preferred).

### CODE REVIEW

**Comment by tkf** [comment]
> Also fix one place where we were acquiring require_lock twice (which is ok since it's reentrant, but not preferred).

Isn't the strategy you've implemented in this PR actually relies on this? IIUC, the intention of `run_package_callbacks` is to release the `require_lock` "completely" so that `require_lock` is not acquired by the time the callbacks are called.

File: test/precompile.jl [hunk 1]
```diff
@@ -767,6 +767,27 @@ precompile_test_harness("package_callbacks") do dir
...
+                     end)
+    Test4_module = :Teste4095a84
+    write(joinpath(dir, "$(Test4_module).jl"),
+          """
+          module $(Test4_module)
+          end
+          """)
+    Base.compilecache(Base.PkgId("$(Test4_module)"))
+    push!(Base.package_callbacks, _->(notify(E); lock(L) do; end))
+    # should not hang here
```
Comment (vtjnash) [inline test/precompile.jl:784]: we perhaps should have an assert that the lock state is unlocked, so that it cannot hang?

**Comment by JeffBezanson** [comment]
> Isn't the strategy you've implemented in this PR actually relies on this?

I don't understand --- if it's locked twice, the `unlock` is not sufficient, so we need it to be locked only once at a time.

**Comment by tkf** [comment]
By "this", I mean the patch introducing `_require_prelocked`. I should've avoided using "this". I meant that, since `run_package_callbacks` requires to turn `require_lock` to unlocked state, we can't lock `require_lock` multiple times in `require` methods. Without `_require_prelocked`, the call chain like `require(into::Module, mod::Symbol)` -> `require(uuidkey::PkgId)` -> `run_package_callbacks` would have called the callbacks while `require_lock` is still being locked.

To put it in another way, your comment in the OP sounded like the change introducing `_require_prelocked` was optional. I thought it was a requirement.

**Pull request merged**